Add `Changes to osu!taiko star rating and performance points` news post (#7915)
ppy/osu-wiki: news/2022-09-27-changes-to-osu-taiko-sr-and-pp.md

In addition to the above, tailored and highly particular nerfs targeted at specific patterns were implemented using code described best as a "cheese detector", which is something that we have since opted to do away with.

The core policy for the all the new reworks under our wing was formed:

**Never assume a specific playstyle or arbitrarily nerf patterns without a strong justification.**

#### The new stamina system

In this section, the keybinds relating to the two colours found within osu!taiko are named k<sup>1</sup>, k<sup>2</sup> and d<sup>1</sup>, d<sup>2</sup>.

Instead of assuming a full-alternate playstyle, we assume that **each note of the same colour is always alternated**. For example, for the pattern kddkddk, we assume the kats will be hit by **k<sup>1</sup>** d<sup>1</sup> d<sup>2</sup> **k<sup>2</sup>** d<sup>1</sup> d<sup>2</sup> **k<sup>1</sup>**, and the dons will be hit by k<sup>1</sup> **d<sup>1</sup> d<sup>2</sup>** k<sup>2</sup> **d<sup>1</sup> d<sup>2</sup>** k<sup>2</sup>.

> The reason for this is to assume the best-case scenario for repeated keystrokes regardless of playstyle. In other words, we aim to approximate the theoretical minimum stamina required to play a given pattern, with any theoretically rollable pattern being rolled.

Reviewer: just going to mention that as a non-taiko player, I have no idea what "rolling" means. But I'm probably not the intended audience for this newspost.

Reviewer 2: My uneducated guess would be pressing three/four keys in a row from left to right or right to left.

Reviewer 3: asked in a taiko discord and got that it's "using the same hand more than once in a row", as opposed to alternating hands.

To combat this, misses will now significantly affect difficulty pp on maps with a max combo below 1000, scaling the miss count by a multiplier of `1000 / total successful hits`. This may significantly impact plays on shorter maps with a higher miss count, as visualised below.

![Effective miss count graph](/wiki/shared/news/2022-09-27-changes-to-osu-taiko-star-rating-and-performance-points/effective-miss-count.jpg)

### Changes to difficulty performance point values

The following changes should result in impressive scores on harder beatmaps rewarding more appropriate performance point values. Please note that these changes only affect the difficulty part of performance, which correlates with the star rating value.

Difficulty pp has been adjusted to hugely nerf previously overweighted low-accuracy scores. In technical terms, difficulty pp scales by the square of the accuracy now instead of linearly, as visualised below:

![Accuracy scaling graph](/wiki/shared/news/2022-09-27-changes-to-osu-taiko-star-rating-and-performance-points/difficulty-accuracy-scaling.jpg)

> To best describe the impact of this change, top player [syaron105](https://osu.ppy.sh/users/8741695) has gained over 6400 total pp because of it.

Author: mmm, it changes scaling significantly but i think it's the wrong supplementing to it, either way the scaling is changed and the general consensus is major buffs to the sr counterpart, not explicitly the accuracy scaling part

> To best describe the impact of this change, top player [syaron105](https://osu.ppy.sh/users/8741695) has gained over 6400 total pp because of it.

Reviewer 2: Can you try rewording what you just said? I can't understand it and I'm not sure if you want changes applied to the text.

> To best describe the impact of this change, top player [syaron105](https://osu.ppy.sh/users/8741695) has gained over 6400 total pp because of it.

Reviewer 3: Fractured provided a [good explanation in the `#difficulty-taiko` channel](https://discord.com/channels/188630481301012481/542532479626772500/1024406176529645590), will apply that

As such, we have opted to take a more modular approach by introducing new changes incrementally, in the hopes of being able to fix any reported issues one step at a time. Moving forward, osu!taiko star rating will be receiving continuous updates, potentially placing star rating and performance points in a state of flux, especially over the coming few months. 

During this time, we will be working on other difficulty components (both small and large-scale) in preparation for future releases. It is our hope that this will continue to help the algorithm reflect the latest trends and expectation in difficulty calculation, but we also apologise in any advance for any disruption this may cause as we begin working things out.

In the star rating part of this rework, we have essentially completely reworked the stamina and colour skills, introducing new concepts and the infrastructure required to support them.

> Performance points have seen significant changes to mod multipliers and a broad change to how accuracy is calculated. Please note however that these changes are *experimental* and did not receive as much scrutiny as the rest of the rework. Consider leaving feedback if you find anything that feels especially unusual.

Reviewer 2: This seems to be in the wrong place. Move it down to be part of the performance section?

Reviewer 2: Also I dunno about "experimental". LIke that sounds like nothing has been reviewed and we're just going gung-ho. Hopefully that's not the case, because it seems amateur.

Reviewer 3: just like star rating, this introduction section also introduces the pp changes, so it's probably in the right place,

Without further ado, let's get into the nitty-gritty of these changes.

## Star rating changes

### Readdressing stamina

#### Issues within the old system

The old system assumed a 2-finger kddk full alternating playstyle (i.e. one in which each subsequent hit is always hit by the other hand regardless of colour). This was implemented as two skills — one for the left hand, and one for the right hand. Instead of using note duration for the split skills, it used the duration between two notes that were pressed with the same hand, a term that we describe as note-pair duration.

> In addition to the above, tailored and highly particular nerfs targeted at specific patterns were implemented using code described best as a "cheese detector", which is something that we have since opted to avoid. 

Reviewer 2: ```suggestion
In addition to the above, tailored and highly particular nerfs targeted at specific patterns were implemented using code described best as a "cheese detector", which is something that we have since opted to do away with.
```

> As such, one core policy for the all the new reworks under our wing was formed:

Reviewer 2: ```suggestion
The core policy for the all the new reworks under our wing was formed:
```
Move below the next header. It's not related to the old system.

#### The new stamina system

In this section, the keybinds relating to the two colours found within osu!taiko are named k<sup>1</sup>, k<sup>2</sup> and d<sup>1</sup>, d<sup>2</sup>.

Instead of assuming a full-alternate playstyle, we assume that **each note of the same colour is always alternated**. For example, for the pattern kddkddk, we assume the kats will be hit by **k<sup>1</sup>** d<sup>1</sup> d<sup>2</sup> **k<sup>2</sup>** d<sup>1</sup> d<sup>2</sup> **k<sup>1</sup>**, and the dons will be hit by k<sup>1</sup> **d<sup>1</sup> d<sup>2</sup>** k<sup>2</sup> **d<sup>1</sup> d<sup>2</sup>** k<sup>2</sup>.

The reason for this is to assume the best-case scenario for repeated keystrokes regardless of playstyle. In other words, we aim to approximate the theoretical minimum stamina required to play a given pattern with any playstyle, with any theoretically rollable pattern being rolled. 

We can't (feasibly), nor is it desirable to, detect playstyles, so we assume the case that requires the least physical repetition, which is to switch keys each time a colour is to be hit, and as such all other existing playstyles will be weaker in comparison to the one we chose.

> ### Readdressing colour

Reviewer 2: ```suggestion
### The "Colour" skill
```

#### Issues within the old system

> Previously, the old system detected only the frequency of changes between colours and had specific penalties for particular repeated patterns. This led to complex patterns containing large quantities of single-coloured notes not being considered properly in regards to overall strain values. 

Reviewer 2: ```suggestion
The previous system only detected the frequency of changes between colours and had specific penalties for particular repeated patterns. This led to complex patterns containing large quantities of single-coloured notes not being considered properly in regards to overall strain values. 
```

> In addition, a penalty was applied to repeat instances of these patterns, similar to how stamina was calculated. This only considered up to five recent occurrences, but if any two or more of such repeats occurred, a harsh penalty was applied.

Reviewer 2: This was just mentioned in the last paragraph..?

#### The new colour system

To summarize shortly, the new colour system is more aware of *all* kinds of patterning used in osu!taiko mapping and is far more capable of formulating appropriate difficulty values from them compared to the old system.

In general, the harder it is to describe a pattern, the harder the pattern will generally be considered. Additionally, the more dense a pattern (or color sequence, as described by the code) is, the harder the algorithm will consider it to be. Patterns that repeat with longer segments between them are also considered more difficult than patterns that repeat closely to one another.

The code and concepts utilized to achieve this are quite extensive. If this sort of thing interests you, feel free to read more in the rework document listed below.

### Calculating the final result

> Given the three skills briefly described above (stamina, rhythm and colour), their *strain* values have now become the number that approximates star rating. 

Reviewer 2: Rhythm is never mentioned above.

Switching from the previous system, where strain and individual skill values formed star rating, strain is now the only summative factor in this calculation via the new *peak* skill.

*Peak* difficulty splits the map into sections and combines the highest strains of each skill in each section, aiming to catch out particularly demanding moments. The final star rating is a weighted sum of the peak skill sections, including a slight multiplier to scale better.

> While this is all much more mathematically complex than described here, you can consult the [osu!taiko rework document](https://docs.google.com/document/d/1Z5GC4DMqOVzeIERMSK3qpQaqjq-sVnhbuoxAwy9qxDs/edit), or dive into the code over in the [osu!(lazer) GitHub repository](https://github.com/ppy/osu) for the full details.

Reviewer 2: ```suggestion
You can consult the [osu!taiko rework document](https://docs.google.com/document/d/1Z5GC4DMqOVzeIERMSK3qpQaqjq-sVnhbuoxAwy9qxDs/edit), or dive into the code over in the [osu!(lazer) GitHub repository](https://github.com/ppy/osu) for the full details on how the calculation works.
```

Reviewer 2: Shgould probably not use "you"

> Previous approaches aimed to scale maps in regards to the [difficulty naming hierarchy](/wiki/Ranking_Criteria/Difficulty_naming), while this rework has deviated away from that goal, aiming instead to reflect true difficulty regardless of naming practices. 

Reviewer 2: aimed at the goal to aim the goal preflection tgisao gl

Reviewer 2: aka needs to be rewritten. and less aiming the goal to aim the goal.

> This also means that star rating values for maps with near-FCs are no longer near 9.5 stars, but rather can expand well into the 10 star range (medal hunters rejoice!).

Reviewer 2: the what? how does star rating have anything to do with FC? confuse.

Reviewer 3: it means that maps that tend to get many near-FCs on their leaderboards now have increased star ratings compared to before

Reviewer 2: I think it should be described as "beatmaps which rarely see FCs achieved" if that's what it's intending to say.

Reviewer 2: Should be worded as such then.

```suggestion
This also means that star rating values for beatmaps where top end players are almost – but not quite – getting full combo scores will now extent well into the 10 star range, rather than capping out at around 9.5 stars (medal hunters rejoice!).
```

## Changes to performance points

> In recent years, osu!taiko has lagged behind the other three game modes in terms of raw performance point values, to the tune of being roughly 40% lower overall. In addition, the survey indicated a great preference for re-evaluating the way performance points are calculated.

Reviewer 2: "lagged behind"? you mean the scale has been off? or time wise? development wise? confuse.

Reviewer 4: the increase of overall top-end pp values for osu!taiko has been behind that of other gamemodes, is what it means

Reviewer 2: Maybe a better way to phrase it is "on a different scale" or something? Is the goal here to bring them roughly into scale with expectations of osu! mode scores?

Previously, performance points in osu!taiko were calculated by summing the "accuracy pp" and "difficulty pp" components of a score together. The following changes should serve to increase these values, make them more relevant to the apparent difficulty of a given map, and improve balance not only between difficulty-increasing mods, but also their difficulty-reducing counterparts.

All of these changes aim to continue the goals outlined in the new star rating rework.

### Globally applied changes

The final step of performance points calculation involves applying a simple multiplier to the calculated value, which consists of both difficulty and accuracy performance points. 

> This allows for simple balancing changes, and works alongside specific multipliers found within their respective separate difficulty values. To begin with, we have made the following changes:

Reviewer 2: This whole text isn't really adding anthing for me.

- The Hidden multiplier has been decreased from **1.10x** to **1.075x**, as a further bonus was placed in difficulty pp, rather than globally.
- Easy now applies a **0.975x** multiplier to help address its problematic nature, especially when paired with DoubleTime.
- The NoFail multiplier has been removed. osu!taiko uses a different HP system compared to osu!, negating the need for any kind of penalty. This also helps prevent "sandbagging" in lower ranks.

Additionally, the community survey highlighted issues with short maps being either underweighted at a full combo, or overweighted on plays with several misses.

To combat this, misses will now significantly affect difficulty pp on maps with a max combo below 1000, scaling the miss count by a multiplier of `1000 / total successful hits`. This may significantly impact plays on shorter maps with a higher miss count, as visualized below.

![Effective miss count graph](/wiki/shared/news/2022-09-27-changes-to-osu-taiko-star-rating-and-performance-points/effective-miss-count.jpg)

### Changes to difficulty performance points values

> The following changes should result in impressive scores on harder beatmaps rewarding more appropriate performance point values, with top player [syaron105](https://osu.ppy.sh/users/8741695) gaining over 6400 total pp!

Reviewer 2: I'm confused. All the points below are small changes (mostly nerf) so how did a user gain 6400 pp as a result? Something is missing in the explanation.

Reviewer 4: i'll rejiggle this around to make more sense

  
While difficulty pp is straightforward, accuracy pp contained a plethora of issues, both to do with reading-based mods (Hidden and Flashlight) and mods that affect the hit windows of Greats (300s), i.e. Easy, DoubleTime and HardRock.

- The impact of the size of the Great (300) hit window on accuracy pp has been reduced significantly. This makes pp less dependent on the 'overall difficulty' value of a map as set in the editor, which greatly reduces the issues of low star rating maps with tight hit windows awarding too much pp.
- Star rating now affects accuracy pp, which replaces most of the previous scaling from the Great hit window. This further supplements the nerfs in accuracy weighting on low star rating scores while buffing high ones, and thus significantly limits abuse of low star rating 3-mod/4-mod scores.
- An additional Hidden + Flashlight bonus has been added to accuracy pp. As this mod combination requires total memorisation of maps, the bonus scales with map length, rewarding longer maps more as visualized below.

![HDFL multiplier graph](/wiki/shared/news/2022-09-27-changes-to-osu-taiko-star-rating-and-performance-points/accuracy-mod-bonus.jpg)
  
## Next steps
  

> Future changes will involve rhythm (which hasn't been touched in these changes so far), and the introduction of  reading, an entirely new concept to osu!taiko performance points calculation, as well as many other large and small-scale changes. 

Reviewer 2: ```suggestion
Future changes will involve rhythm (which hasn't been touched in these changes so far), and the introduction of a "reading" skill, an entirely new concept to osu!taiko performance points calculation, as well as many other large and small-scale changes. 
```

This time, we have opted to take a more modular approach by introducing new changes incrementally, in the hopes of being able to fix any reported issues one step at a time. Moving forward, osu!taiko star rating will be receiving continuous updates, placing star rating and performance points in a state of flux, especially over the coming few months. 

In the star rating part of this rework, we have focused on "stamina" and "colour" skills.

Performance points have seen significant changes to mod multipliers and a broad change to how accuracy is calculated. Please note however that these changes are *experimental* and did not receive as much scrutiny as the rest of the rework. Consider leaving feedback if you find anything that feels especially unusual.

Without further ado, let's get into the nitty-gritty of these changes.

## Star rating changes

> ### "Stamina" skill

Reviewer 2: This should probably have added a "The" before, to match the header below (or remove the "The" below, whatever feels better to you).
